src/runtime_src/core/tools/xbflash2/OO_Dump_Qspips.cpp

```diff
@@ -0,0 +1,198 @@
+
+namespace {
+
+int
+qspips_readback(po::variables_map& vm)
+{
+    size_t offset = 0, len = FLASH_SIZE;
+    std::string bdf;
+    std::string soffset;
+    std::string slength;
+    std::string flash_type;
+    std::string output;
+    std::string sBar, sBarOffset;
+    size_t baroff = INVALID_OFFSET;
+    int bar = 0;
+
+    XBU::sudo_or_throw_err();
+
+    try {
+        if (vm.count("device")) {
+            bdf = vm["device"].as<std::string>();
+        }
+        else {
+            std::cerr << "\nERROR: Device not specified. Please specify a single device using --device option\n";
+            throw std::errc::invalid_argument;
+        }
+
+        if (vm.count("output")) {
+            output = vm["output"].as<std::string>();
+        }
+        else {
+            std::cerr << "\nERROR: output not specified. Please specify the output file path using --output option\n";
+            throw std::errc::invalid_argument;
+        }
+    }
+    catch (...) {
+        throw std::errc::invalid_argument;
+    }
+
+    try {
+        //Optional arguments
+        if (vm.count("flash-part"))
+            flash_type = vm["flash-part"].as<std::string>();
+
+        if (vm.count("offset")) {
+            soffset = vm["offset"].as<std::string>();
+            if (!soffset.empty()) {
+                std::stringstream sstream(soffset);
+                sstream >> offset;
+            }
+        }
+
+        if (vm.count("length")) {
+            slength = vm["length"].as<std::string>();
+            if (!slength.empty()) {
+                std::stringstream sstream(slength);
+                sstream >> len;
+            }
+        }
+    }
+    catch (...) {
+    }
+
+    //Optional arguments
+    try {
+        if (vm.count("bar")) {
+            sBar = vm["bar"].as<std::string>();
+            if (!sBar.empty())
+                bar = std::stoi(sBar);
+        }
+        if (vm.count("bar-offset")) {
+            sBarOffset = vm["bar-offset"].as<std::string>();
+            if (!sBarOffset.empty()) {
+                std::stringstream sstream(sBarOffset);
+                sstream >> baroff;
+            }
+        }
+    }
+    catch (...) {
+    }
+
+    if (bdf.empty() || output.empty())
+        throw std::errc::invalid_argument;
+
+    std::cout << "Read out flash"
+        << boost::format("[0x%x, 0x%x] on device %s to %s\n") % offset % (offset + len) % bdf % output;
+
+    pcidev::pci_device dev(bdf, bar, baroff, flash_type);
+    XQSPIPS_Flasher qspips(&dev);
+
+    return qspips.xclReadBack(output, offset, len);
+}
+} //end namespace 
+
+// ----- C L A S S   M E T H O D S -------------------------------------------
+
+OO_Dump_Qspips::OO_Dump_Qspips( const std::string &_longName, bool _isHidden )
+    : OptionOptions(_longName, _isHidden, "Flash type - qspips")
+    , m_help(false)
+{
+  m_optionsDescription.add_options()
+      ("device,d", po::value<std::string>(), "Bus:Device.Function\" (e.g., 0000:d8:00.0) device of interest.\n")
```

Comment:
Is this the order you wish the options to be displayed?

Reply:
Yes @rozumx